src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs

```diff
@@ -0,0 +1,62 @@
+using System.Collections.Generic;
+
+namespace NUnit.Framework.Constraints.Comparers
+{
+    internal struct ComparisonState
+    {
+        /// <summary>
+        /// Flag indicating whether or not this is the top level comparison.
+        /// </summary>
+        public readonly bool TopLevelComparison;
+
+        /// <summary>
+        /// A list of tracked comparisons
+        /// </summary>
+        private readonly Comparison[] _comparisons;
+
+        public ComparisonState(bool topLevelComparison)
+        {
+            TopLevelComparison = topLevelComparison;
+            _comparisons = new Comparison[0];
+        }
+
+        private ComparisonState(bool topLevelComparison, Comparison[] comparisons)
+        {
+            TopLevelComparison = topLevelComparison;
+            _comparisons = comparisons;
+        }
+
+        public ComparisonState PushComparison(bool topLevelComparison, object x, object y)
```

Comment:
You only ever pass in `false` for `topLevelComparison`, so is this needed? Don't we only start with a top-level and all pushes are therefore children?

Reply:
Yes! It seems like it could be calculated as `comparisons.Length == 0` or something like that.

Reply:
Agreed, this is ideal. The only wrinkle to making this calculated is backwards compatibility. This call is downstream from a public API where this boolean can be passed in as an argument:
https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs#L157

Reply:
Oh, but that is a good point. I can omit passing it here, since it would always be false deeper down the stack

Reply:
Done. I ended up defaulting to a hard-coded `false` on push. I'm a little concerned that making this fully calculated could break someone

Reply:
I agree that we don't want to make any behavioral changes and I guess there's no way to avoid it because the top-level flag would sometimes be false even when the stack is empty. We could emulate it by pushing a `new object()` comparison, but that seems less straightforward.